_posts/2021-01-01-getting-started-with-graphql.md

```diff
@@ -0,0 +1,288 @@
+    super();
+    this.manufacturerRepository = manufacturerRepository;
+  }
+
+  public Manufacturer getMadeBy(final Product product) {
+    return manufacturerRepository.getManufacturerById(product.getManufacturerID());
+  }
+}
+```
+This resolver fetches the `Manufacturer` record corresponding to a `Product`.
+
+### Connecting to Datasources and Applying Middleware Logic
+Next, we will enable our resolvers to fetch data from underlying data sources like a database or web service. For this example, we have configured an in-memory H2 database as the data store for `products` and `manufacturers`. We use Spring JDBC to retrieve data from the database and put this logic in separate repository classes. 
+
+Apart from fetching data, we can also build different categories of middleware logic in this business service layer, like authorization of incoming requests, applying filters on data fetched from backend, transformation into backend data models, and also caching any less frequently changing data.
+
+### Running Application
+After compiling and running the application, we can send GraphQL queries to the endpoint http://localhost:8080/graphql. A sample request and response captured in postman tool is shown here :
+
+![Snippet from Postman Tool](/assets/img/posts/graphql-intro/results.png)
+
+## GraphQL vs REST
+REST has been the de-facto style for building APIs. Good API designs are usually driven by consumer needs which can be varied. If we consider an e-commerce site, we might want to show recent purchases of a customer on an order history page and last viewed products on a profile page. 
+
+With REST we will require two different APIs although we are working with the same product data. Alternately we might fetch the entire product data with all its relations every time even though we only need a part of the data. 
+
+GraphQL tries to solve these problems of over fetching or under fetching data. With GraphQL, we will have a single endpoint on which the consumer can send different queries depending on the data of interest. Let us look into how GraphQL differs from REST.
+
+### Shape of the API
+REST API is based on resources that are identified by URLs and an HTTP method (GET, POST, PUT, DELETE.) indicating one of the CRUD operations. GraphQL in contrast is based on a data graph that is returned in response to a request sent as a query to a fixed endpoint.
+
+### HTTP Status Codes
+REST APIs are mostly designed to return 200 series status codes for success and 400 and 500 series for failures. GraphQL APIs return 200 as status code irrespective of whether it is a success or failure. 
+
+### Health Check
+With REST APIs, we check for a 200 status code on a specific endpoint to check if the API is healthy and capable of serving the requests. In GraphQL, health checking is relatively complex since the monitoring function needs to parse the response body to check the server status.
+
+### Caching
```

Comment:
This section hides the fact that common (server-side) caching mechanisms fail with GraphQL. I think that is worth mentioning.